File: LeetCode/Graph/DFS/Python/Leetcode687-longest-univalue-path.py
class Solution:
    def longestUnivaluePath(self, root: Optional[TreeNode]) -> int:
        res = 0
        def dfs(node):
            nonlocal res
            if not node: return 0
            l,r = dfs(node.left),dfs(node.right)
            ll = l + 1 if node.left and node.left.val == node.val else 0
            rr = r + 1 if node.right and node.right.val == node.val else 0
            res = max(res,ll + rr)
            return max(ll,rr)
        dfs(root)
        return res

File: LeetCode/Graph/DFS/Golang/Leetcode687-longest-univalue-path.go
package sbgo

func longestUnivaluePath(root *TreeNode) int {
  res := 0
  var dfs func(node *TreeNode) int 
  max := func(i,j int) int {
    if i < j {
      return j
    }
    return i
  }
  dfs = func(node *TreeNode) int {
    if node == nil {
      return 0
    }
    l,r := dfs(node.Left),dfs(node.Right)
    ll,rr := 0,0
    if node.Left != nil &&  node.Left.Val == node.Val {
      ll = l + 1
    }
    if node.Right != nil && node.Right.Val == node.Val {
      rr = r + 1
    }
    res = max(res,ll + rr)
    return max(ll,rr)
  }
  dfs(root)
  return res
}
